valkey-k8s-operator

```diff
@@ -0,0 +1,267 @@
+
+Motivation
+
+Operating Valkey at scale in Kubernetes environments remains complex and time-consuming. 
+Most enterprise users currently manage Valkey deployments manually or via generic Helm charts, lacking production-readiness, automation, or compliance guarantees.
+
+This Operator addresses these challenges by:
+- Providing a Kubernetes-native way to deploy and manage cluster mode Valkey, which is the only deployment type that provides HA and horizontal scaling.
+- Enabling self-service for developers via CRDs
+- Automating TLS setup, monitoring, and failover
+- Supporting advanced production use cases:
+  - Cluster mode deployments with multiple Sentinel-managed shards
+  - Persistent storage and backup/restore workflows
+- Providing simplified standalone or Sentinel-only deployments for smaller or test environments where full cluster mode is unnecessary
+- Matching enterprise expectations for security, observability, and compliance
+
+Unlike Redis, which has various third-party operators with unclear support models, Valkey has the opportunity to lead with an officially supported Kubernetes-native experience.
+
+Design considerations
+
+The operator will be implemented using Kubebuilder and expose a Custom Resource Definition (CRD) named `Valkey`, allowing users to declaratively deploy Valkey instances via YAML. 
+
+Key design principles:
+- User-first: Apply 10 lines of YAML and get a working, HA Valkey deployment.
+- Cluster-first focus: Prioritize cluster mode as the production-grade setup — HA, scalable, and resilient
+- Fallback / simpler modes: Support standalone and Sentinel HA deployments for small-scale or testing environments
+- Composability: Work alongside tools like cert-manager, Prometheus Operator, and Everest.
+- Security: TLS and credential management must be default-safe.
+- Modularity: Start simple (standalone + Sentinel), then iterate (autoscaling, dashboards).
+- Observability: First-class Prometheus support out-of-the-box.
+- Deployment approach: The operator will rely on either:
+  - A fully open-source Helm chart maintained internally
+  - Direct deployment manifests managed by the operator itself
+
+Comparative notes:
+- SAP’s internal `valkey-operator` is optimized for Sentinel and metrics but is not distribution-friendly.
+- Hyperspike’s operator is more lightweight and public, but lacks Sentinel and TLS support.
+This RFC proposes an open-source Valkey operator that balances ease of use and production-grade features.
+
+Specification
+
+CRD Schema Overview
+mode: cluster (recommended for production), sentinel (HA for small-scale), standalone (simple deployments or dev/test)
+
+```yaml
+apiVersion: cache.percona.com/v1alpha1
+kind: Valkey
+metadata:
+  name: my-valkey
+spec:
+  mode: cluster
+  replicas: 3
+  sentinel:
+    enabled: true
```

Comment:
Cluster mode with sentinel is not possible.

I don't think we should have this yaml code under specification. It looks more like an example to me.